tiledb/sm/query/readers/aggregators/sum_aggregator.h

```diff
@@ -0,0 +1,186 @@
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This file defines class SumAggregator.
+ */
+
+#ifndef TILEDB_SUM_AGGREGATOR_H
+#define TILEDB_SUM_AGGREGATOR_H
+
+#include "tiledb/common/status.h"
+#include "tiledb/sm/enums/layout.h"
+#include "tiledb/sm/query/readers/aggregators/iaggregator.h"
+
+using namespace tiledb::common;
+
+namespace tiledb {
+namespace sm {
+
+#define SUM_TYPE_DATA(T, SUM_T) \
+  template <>                   \
+  struct sum_type_data<T> {     \
+    using type = T;             \
+    typedef SUM_T sum_type;     \
+  };
+
+/** Convert basic type to a sum type. **/
+template <typename T>
+struct sum_type_data;
+
+SUM_TYPE_DATA(int8_t, int64_t);
+SUM_TYPE_DATA(uint8_t, uint64_t);
+SUM_TYPE_DATA(int16_t, int64_t);
+SUM_TYPE_DATA(uint16_t, uint64_t);
+SUM_TYPE_DATA(int32_t, int64_t);
+SUM_TYPE_DATA(uint32_t, uint64_t);
+SUM_TYPE_DATA(int64_t, int64_t);
+SUM_TYPE_DATA(uint64_t, uint64_t);
+SUM_TYPE_DATA(float, double);
+SUM_TYPE_DATA(double, double);
+
+class QueryBuffer;
+
+/**
+ * Sum function that prevent wrap arounds on overflow.
+ *
+ * @param value Value to add to the sum.
+ * @param sum Computed sum.
+ */
+template <typename SUM_T>
+void safe_sum(SUM_T value, SUM_T& sum);
+template <typename T>
+class SumAggregator : public IAggregator {
+ public:
+  /* ********************************* */
+  /*     CONSTRUCTORS & DESTRUCTORS    */
+  /* ********************************* */
+
+  SumAggregator() = delete;
+
+  /**
+   * Constructor.
+   *
+   * @param field_name Name of the field that this aggregate will process.
+   * @param array_schema Array schema.
+   */
+  SumAggregator(const std::string field_name, const ArraySchema& schema);
+
+  DISABLE_COPY_AND_COPY_ASSIGN(SumAggregator);
+  DISABLE_MOVE_AND_MOVE_ASSIGN(SumAggregator);
+
+  /* ********************************* */
+  /*                API                */
+  /* ********************************* */
+
+  /** Returns the field name for the aggregator. */
+  std::string field_name() override {
+    return field_name_;
+  }
+
+  /** Returns if the aggregation is var sized or not. */
+  bool var_sized() override {
+    return false;
+  };
+
+  /** Returns if the aggregate needs to be recomputed on overflow. */
+  bool need_recompute_on_overflow() override {
+    return true;
```

Comment:
Oh, this is what overflows? I'm not at all seeing how if we overflowed once we'd be able to recompute the sum again without also overflowing? Or is this a "we'll return partial query results to avoid overflows" sort of behavior?

That seems odd to me on the face of it from a data consumption point of view. I wonder what happens in SQL land if a sum overflows? Or do they just swap to a bignum implementation?

Reply:
Now if we talk about integer overflow, once we overflowed the value, we don't do any switch to a larger container... We'll just return max uint64, int64 or double.